src/LayoutManager/LayoutManager.ts

```diff
@@ -0,0 +1,294 @@
+    } else if (context.type === 'removed') {
+      context.targets.forEach((object) => this.unsubscribe(context, object));
+    }
+
+    //  fire layout hook and event (event will fire only for layouts after initialization layout)
+    target.onBeforeLayout({
+      context,
+    });
+    target.fire('layout:before', {
+      context,
+    });
+
+    if (context.type === 'imperative' && context.deep) {
+      const { strategy: _, ...tricklingContext } = context;
+      // traverse the tree
+      target.forEachObject((object) => {
+        (object as Group).layoutManager?.performLayout({
+          ...tricklingContext,
+          bubbles: false,
+          target: object as Group,
+        });
+      });
+    }
+  }
+
+  protected getLayoutResult(
+    context: StrictLayoutContext
+  ): Required<LayoutResult> | undefined {
+    const { target } = context;
+    const prevCenter =
+      context.type === 'initialization'
+        ? new Point()
+        : target.getRelativeCenterPoint();
+    const result = context.strategy.calcLayoutResult(
+      context,
+      target.getObjects()
+    );
+    if (!result) {
+      return;
+    }
+    const nextCenter = new Point(result.centerX, result.centerY);
+    const correction = new Point(
+      result.correctionX ?? 0,
+      result.correctionY ?? 0
+    );
+    const relativeCorrection = new Point(
+      result.relativeCorrectionX ?? 0,
+      result.relativeCorrectionY ?? 0
+    );
+    const offset =
+      context.type === 'initialization' && context.objectsRelativeToGroup
+        ? new Point()
+        : prevCenter
+            .subtract(nextCenter)
+            .add(correction)
+            .transform(
+              // in `initialization` we do not account for target's transformation matrix
+              context.type === 'initialization'
+                ? iMatrix
+                : invertTransform(target.calcOwnMatrix()),
+              true
+            )
+            .add(relativeCorrection);
+    return {
+      result,
+      prevCenter,
+      nextCenter,
+      offset,
+    };
+  }
+
+  protected commitLayout(
+    context: StrictLayoutContext,
+    layoutResult: Required<LayoutResult>
+  ) {
+    const { target } = context;
+    const {
+      result: { width, height },
+      nextCenter,
+    } = layoutResult;
+    // set dimensions
+    target.set({ width, height });
+    // layout descendants
+    this.layoutObjects(context, layoutResult);
+    //  set position
+    // in `initialization` we do not account for target's transformation matrix
+    if (context.type === 'initialization') {
+      // TODO: what about strokeWidth?
+      const origin = nextCenter.add(
+        new Point(width, height).multiply(
+          new Point(
+            resolveOrigin(target.originX),
+            resolveOrigin(target.originY)
+          )
+        )
+      );
+      target.set({ left: context.x ?? origin.x, top: context.y ?? origin.y });
```

Comment:
```suggestion
      const left = context.x ?? nextCenter.x + width * resolveOrigin(target.originX);
      const top = context.y ?? nextCenter.y + height * resolveOrigin(target.originY);
      target.set({ left, top });
```
is more concise, calculates only what we need and in a less expensive way and is equally readable

Reply:
I dislike it but OK